skimage/segmentation/morphsnakes.py

```diff
@@ -0,0 +1,442 @@
+# -*- coding: utf-8 -*-
+
+from itertools import cycle
+
+import numpy as np
+from scipy import ndimage as ndi
+
+from .._shared.utils import assert_nD
+
+__all__ = ['morphological_chan_vese',
+           'morphological_geodesic_active_contour',
+           'inverse_gaussian_gradient',
+           'circle_level_set',
+           'checkerboard_level_set'
+           ]
+
+
+class _fcycle(object):
+
+    def __init__(self, iterable):
+        """Call functions from the iterable each time it is called."""
+        self.funcs = cycle(iterable)
+
+    def __call__(self, *args, **kwargs):
+        f = next(self.funcs)
+        return f(*args, **kwargs)
+
+
+# SI and IS operators for 2D and 3D.
+_P2 = [np.eye(3),
+       np.array([[0, 1, 0]] * 3),
+       np.flipud(np.eye(3)),
+       np.rot90([[0, 1, 0]] * 3)]
+_P3 = [np.zeros((3, 3, 3)) for i in range(9)]
+
+_P3[0][:, :, 1] = 1
+_P3[1][:, 1, :] = 1
+_P3[2][1, :, :] = 1
+_P3[3][:, [0, 1, 2], [0, 1, 2]] = 1
+_P3[4][:, [0, 1, 2], [2, 1, 0]] = 1
+_P3[5][[0, 1, 2], :, [0, 1, 2]] = 1
+_P3[6][[0, 1, 2], :, [2, 1, 0]] = 1
+_P3[7][[0, 1, 2], [0, 1, 2], :] = 1
+_P3[8][[0, 1, 2], [2, 1, 0], :] = 1
+
+
+def sup_inf(u):
+    """SI operator."""
+
+    if np.ndim(u) == 2:
+        P = _P2
+    elif np.ndim(u) == 3:
+        P = _P3
+    else:
+        raise ValueError("u has an invalid number of dimensions "
+                         "(should be 2 or 3)")
+
+    erosions = []
+    for P_i in P:
+        erosions.append(ndi.binary_erosion(u, P_i))
+
+    return np.array(erosions, dtype=np.int8).max(0)
+
+
+def inf_sup(u):
+    """IS operator."""
+
+    if np.ndim(u) == 2:
+        P = _P2
+    elif np.ndim(u) == 3:
+        P = _P3
+    else:
+        raise ValueError("u has an invalid number of dimensions "
+                         "(should be 2 or 3)")
+
+    dilations = []
+    for P_i in P:
+        dilations.append(ndi.binary_dilation(u, P_i))
+
+    return np.array(dilations, dtype=np.int8).min(0)
+
+
+_curvop = _fcycle([lambda u: sup_inf(inf_sup(u)),   # SIoIS
+                   lambda u: inf_sup(sup_inf(u))])  # ISoSI
+
+
+def _check_input(image, init_level_set):
+    """Check that shapes of `image` and `init_level_set` match."""
+    assert_nD(image, [2, 3])
+
+    if len(image.shape) != len(init_level_set.shape):
+        raise ValueError("The dimensions of the initial level set do not "
+                         "match the dimensions of the image.")
+
+
+def _init_level_set(init_level_set, image_shape):
```

Comment:
`level_set`s are partially duplicated by those in `chan_vese`. We should consider factoring them out.